File: lib/shared/bucketing-assembly-script/assembly-script-src/bucketing/bucketing.ts
import { JSON } from 'assemblyscript-json'
import { first, last } from '../helpers/lodashHelpers'
import {
    ConfigBody, Target as PublicTarget, Feature as PublicFeature, BucketedUserConfig,
    Rollout as PublicRollout, DVCPopulatedUser, SDKVariable, SDKFeature, RolloutStage,
    Target, Variation, Variable
} from '../types'

// TODO: implement murmurhash logic in AS
// import murmurhash from 'murmurhash'
import { evaluateOperator } from './segmentation'

// Max value of an unsigned 32-bit integer, which is what murmurhash returns
const MAX_HASH_VALUE: i32 = 4294967295
const baseSeed: i32 = 1

class BoundedHash {
    public rolloutHash: i32
    public bucketingHash: i32
}

export function generateBoundedHashes(user_id: string, target_id: string): BoundedHash {
    // The seed provided to murmurhash must be a number
    // So we first hash the target_id with a constant seed

    const targetHash = 1 //murmurhash.v3(target_id, baseSeed)
    const boundedHash: BoundedHash = {
        rolloutHash: generateBoundedHash(user_id + '_rollout', targetHash),
        bucketingHash: generateBoundedHash(user_id, targetHash)
    }
    return boundedHash
}

export function generateBoundedHash(input: string, hashSeed: i32): i32 {
    const hash = 1 //murmurhash.v3(input, hashSeed)
    return hash / MAX_HASH_VALUE
}

/**
 * Given the feature and a hash of the user_id, bucket the user according to the variation distribution percentages
 */
export function decideTargetVariation(target: PublicTarget, boundedHash: i32): string {
    //TODO: figure out sorting
    const variations = target.distribution
        //.sort((a, b) => a._variation > b._variation)

    let distributionIndex: f64 = 0
    const previousDistributionIndex: f64 = 0
    for (let i = 0; i < variations.length; i++) {
        const variation = variations[i]
        distributionIndex += variation.percentage
        if (boundedHash >= previousDistributionIndex && boundedHash < distributionIndex) {
            return variation._variation
        }
    }
    throw new Error('Failed to decide target variation')
}

export function getCurrentRolloutPercentage(rollout: PublicRollout, currentDate: Date): f64 {
    const start = rollout.startPercentage
    const startDateTime = rollout.startDate.getTime()
    const currentDateTime = currentDate.getTime()

    if (rollout.type === 'schedule') {
        return currentDateTime >= startDateTime ? 1 : 0
    }

    if (!rollout.stages) {
        return 0
    }
    const stages = rollout.stages as RolloutStage[]
    const currentStages: RolloutStage[] = []
    const nextStages: RolloutStage[] = []
    for (let i = 0; i < stages.length; i++) {
        const stage = stages[i]
        const stageTime = stage.date.getTime()
        if (stageTime <= currentDateTime) {
            currentStages.push(stage)
        } else {
            nextStages.push(stage)
        }
    }

    const _currentStage = last(currentStages)
    const nextStage = first(nextStages)

    let currentStage = _currentStage
    if (!_currentStage && (startDateTime < currentDateTime)) {
        const jsonObj = new JSON.Obj()
        jsonObj.set('type', '')
        jsonObj.set('percentage', start)
        jsonObj.set('date', rollout.startDate.toISOString())
        currentStage = new RolloutStage(jsonObj)
    }

    if (!currentStage) {
        return 0
    }

    if (!nextStage || nextStage.type === 'discrete') {
        return currentStage.percentage
    }

    const currentDatePercentage = (currentDateTime - currentStage.date.getTime()) /
            (nextStage.date.getTime() - currentStage.date.getTime())

    if (currentDatePercentage === 0) {
        return 0
    }

    return (
        (currentStage.percentage + (nextStage.percentage - currentStage.percentage))
        * f64(currentDatePercentage)
    )
}

export function doesUserPassRollout(rollout: PublicRollout | null, boundedHash: i32): bool {
    if (!rollout) return true

    const rolloutPercentage = getCurrentRolloutPercentage(rollout, new Date(Date.now()))
    return !!rolloutPercentage && (boundedHash <= rolloutPercentage)
}

export function bucketForSegmentedFeature(boundedHash: i32, target: PublicTarget): string {
    return decideTargetVariation(target, boundedHash)
}

class SegmentedFeatureData {
    public feature: PublicFeature
    public target: PublicTarget
}

export function getSegmentedFeatureDataFromConfig(
    config: ConfigBody,
    user: DVCPopulatedUser
): SegmentedFeatureData[] {
    const accumulator: SegmentedFeatureData[] = []

    for (let y = 0; y < config.features.length; y++) {
        const feature = config.features[y]

        // Returns the first target for which the user passes segmentation
        let segmentedFeatureTarget: Target | null = null
        for (let i = 0; i < feature.configuration.targets.length; i++) {
            const target = feature.configuration.targets[i]
            if (evaluateOperator(target._audience.filters, user)) {
                segmentedFeatureTarget = target
                break
            }
        }

        if (segmentedFeatureTarget) {
            const featureData: SegmentedFeatureData = {
                feature,
                target: segmentedFeatureTarget
            }
            accumulator.push(featureData)
        }
    }
    return accumulator
}

export function generateKnownVariableKeys(
    variableHashes: Map<string, i64>,
    variableMap: Map<string, SDKVariable>
): i64[] {
    const knownVariableKeys: i64[] = []
    const hashKeys = variableHashes.keys()
    for (let i = 0; i < hashKeys.length; i++) {
        const key = hashKeys[i]
        const hash = variableHashes.get(key)
        const variable = variableMap.get(key)
        if (!variable) {
            knownVariableKeys.push(hash)
        }
    }
    return knownVariableKeys
}

export function generateBucketedConfig(
    config: ConfigBody,
    user: DVCPopulatedUser
): BucketedUserConfig {
    const variableMap = new Map<string, SDKVariable>()
    const featureKeyMap = new Map<string, SDKFeature>()
    const featureVariationMap = new Map<string, string>()
    const segmentedFeatures = getSegmentedFeatureDataFromConfig(config, user)

    for (let i = 0; i < segmentedFeatures.length; i++) {
        const segmentedFeaturesData = segmentedFeatures[i]
        const feature = segmentedFeaturesData.feature
        const target = segmentedFeaturesData.target

        const boundedHashData = generateBoundedHashes(user.user_id, target._id)
        const rolloutHash = boundedHashData.rolloutHash
        const bucketingHash = boundedHashData.bucketingHash
        if (target.rollout && !doesUserPassRollout(target.rollout, rolloutHash)) {
            continue
        }

        const variation_id = bucketForSegmentedFeature(bucketingHash, target)
        featureKeyMap.set(feature.key, new SDKFeature(
            feature._id,
            feature.key,
            feature.type,
            variation_id,
            null
        ))
        featureVariationMap.set(feature._id, variation_id)

        let variation: Variation | null = null
        for (let i = 0; i < feature.variations.length; i++) {
            const featVariation = feature.variations[i]
            if (featVariation._id === variation_id) {
                variation = featVariation
                break
            }
        }
        if (!variation) {
            throw new Error(`Config missing variation: ${variation_id}`)
        }

        for (let y = 0; y < variation.variables.length; y++) {
            const variationVar = variation.variables[y]

            // Find variable
            let variable: Variable | null = null
            for (let u = 0; u < config.variables.length; u++) {
                const configVar = config.variables[u]
                if (configVar._id === variationVar._var) {
                    variable = configVar
                    break
                }
            }
            if (!variable) {
                throw new Error(`Config missing variable: ${variationVar._var}`)
            }

            const newVar = new SDKVariable(
                variable._id,
                variable.type,
                variable.key,
                variationVar.value,
                null
            )
            variableMap.set(variable.key, newVar)
        }
    }

    return new BucketedUserConfig(
        config.project,
        config.environment,
        featureKeyMap,
        featureVariationMap,
        variableMap,
        generateKnownVariableKeys(config.variableHashes, variableMap)
    )
}

File: lib/shared/bucketing-zig/src/bucketing.zig
const std = @import("std");
const types = @import("types.zig");
const Allocator = std.mem.Allocator;

// Re-export key types
const BoundedHash = types.BoundedHash;
const DVCPopulatedUser = types.DVCPopulatedUser;
const ConfigBody = types.ConfigBody;
const CustomData = types.CustomData;
const BucketedUserConfig = types.BucketedUserConfig;
const BucketedVariableResponse = types.BucketedVariableResponse;
const SDKVariable = types.SDKVariable;
const VariableType = types.VariableType;
const EvalReason = types.EvalReason;
const JsonValue = types.JsonValue;
const Project = types.Project;
const Environment = types.Environment;

// Constants
const BASE_SEED: u32 = 0x12345678;
const MAX_HASH_VALUE: f64 = 4294967295.0; // 2^32 - 1

/// MurmurHash3 implementation for 32-bit hashes
pub fn murmurhash3(key: []const u8, seed: u32) u32 {
    const c1: u32 = 0xcc9e2d51;
    const c2: u32 = 0x1b873593;
    const r1: u32 = 15;
    const r2: u32 = 13;
    const m: u32 = 5;
    const n: u32 = 0xe6546b64;

    const len = key.len;
    var hash = seed;
    var i: usize = 0;

    // Process 4-byte chunks
    while (i + 4 <= len) {
        var k: u32 = 0;
        k |= @as(u32, key[i]);
        k |= @as(u32, key[i + 1]) << 8;
        k |= @as(u32, key[i + 2]) << 16;
        k |= @as(u32, key[i + 3]) << 24;

        k = std.math.rotl(u32, k *% c1, r1) *% c2;
        hash ^= k;
        hash = std.math.rotl(u32, hash, r2) *% m +% n;

        i += 4;
    }

    // Process remaining bytes
    var k1: u32 = 0;
    switch (len & 3) {
        3 => {
            k1 ^= @as(u32, key[i + 2]) << 16;
            k1 ^= @as(u32, key[i + 1]) << 8;
            k1 ^= @as(u32, key[i]);
        },
        2 => {
            k1 ^= @as(u32, key[i + 1]) << 8;
            k1 ^= @as(u32, key[i]);
        },
        1 => {
            k1 ^= @as(u32, key[i]);
        },
        else => {},
    }

    if (k1 != 0) {
        k1 = std.math.rotl(u32, k1 *% c1, r1) *% c2;
        hash ^= k1;
    }

    // Finalization
    hash ^= @as(u32, @intCast(len));
    hash ^= hash >> 16;
    hash = hash *% 0x85ebca6b;
    hash ^= hash >> 13;
    hash = hash *% 0xc2b2ae35;
    hash ^= hash >> 16;

    return hash;
}

/// Generate a bounded hash value between 0 and 1
pub fn generateBoundedHashValue(input: []const u8, hash_seed: u32) f64 {
    const hash = murmurhash3(input, hash_seed);
    return @as(f64, @floatFromInt(hash)) / MAX_HASH_VALUE;
}

/// Generate bounded hashes for rollout and bucketing
pub fn generateBoundedHashes(user_id: []const u8, target_id: []const u8) BoundedHash {
    // Hash the target_id with base seed to get target hash
    const target_hash = murmurhash3(target_id, BASE_SEED);
    
    // Create rollout input by concatenating user_id + "_rollout"
    var rollout_buffer: [256]u8 = undefined;
    const rollout_input = std.fmt.bufPrint(&rollout_buffer, "{s}_rollout", .{user_id}) catch user_id;
    
    const rollout_hash = generateBoundedHashValue(rollout_input, target_hash);
    const bucketing_hash = generateBoundedHashValue(user_id, target_hash);
    
    return BoundedHash{
        .rollout_hash = rollout_hash,
        .bucketing_hash = bucketing_hash,
    };
}

/// Generate bucketed configuration for a user
pub fn generateBucketedConfig(
    config: ConfigBody,
    _: DVCPopulatedUser,
    _: CustomData,
    _: ?CustomData,
    allocator: Allocator
) !BucketedUserConfig {
    // Create a bucketed config based on the user and configuration
    const bucketed_config = types.BucketedUserConfig.init(
        try Project.init(config.project.id, config.project.key, allocator),
        try Environment.init(config.environment.id, config.environment.key, allocator),
        allocator
    );
    
    // For now, return a basic config
    // TODO: Implement full bucketing logic for features and variables
    
    return bucketed_config;
}

/// Generate a bucketed variable for a specific user and variable key
pub fn generateBucketedVariableForUser(
    config: ConfigBody,
    _: DVCPopulatedUser,
    variable_key: []const u8,
    _: CustomData,
    allocator: Allocator
) !?BucketedVariableResponse {
    // Find the variable in the config
    for (config.variables.items) |variable| {
        if (std.mem.eql(u8, variable.key, variable_key)) {
            // Found the variable, now we need to determine its value based on bucketing
            
            // For now, create a basic variable with default values
            const eval_reason = try EvalReason.init("default", "Variable default", "", allocator);
            const default_value = JsonValue{ .string = try allocator.dupe(u8, "default_value") };
            
            const sdk_variable = try SDKVariable.init(
                variable.id,
                variable.type,
                variable.key,
                default_value,
                null, // feature
                eval_reason,
                allocator
            );
            
            // Create a default variation and feature
            const variation = try types.Variation.init("default_variation", "Default", "default", allocator);
            const feature = try types.Feature.init("default_feature", "default_feature_key", "release", allocator);
            
            const response = types.BucketedVariableResponse.init(sdk_variable, variation, feature);
            return response;
        }
    }
    
    // Variable not found
    return null;
}

/// Evaluate if a user passes segmentation for a feature (stub implementation)
pub fn evaluateSegmentationForFeature(
    _: ConfigBody,
    _: types.Feature,
    _: DVCPopulatedUser,
    _: CustomData
) bool {
    // TODO: Implement segmentation evaluation logic
    // For now, always return true
    return true;
}

/// Determine which variation a user should receive for a feature
pub fn bucketUserForVariation(
    feature: types.Feature,
    _: DVCPopulatedUser
) ?types.Variation {
    // TODO: Implement variation bucketing logic
    // For now, return the first variation if available
    if (feature.variations.items.len > 0) {
        return feature.variations.items[0];
    }
    return null;
}

/// Check if user passes rollout percentage
pub fn doesUserPassRollout(
    rollout_percentage: f64,
    bounded_hash: f64
) bool {
    return bounded_hash <= rollout_percentage;
}

/// Get current rollout percentage based on date and rollout configuration
pub fn getCurrentRolloutPercentage(
    start_percentage: f64,
    start_date: i64,
    current_date: i64,
    _: []const u8
) f64 {
    // TODO: Implement full rollout logic with stages
    // For now, return start percentage if current date >= start date
    if (current_date >= start_date) {
        return start_percentage;
    }
    return 0.0;
}

// Test functions for validation
test "murmurhash3 basic functionality" {
    const input = "hello";
    const seed: u32 = 0;
    const hash = murmurhash3(input, seed);
    
    // MurmurHash3 should produce consistent results
    const hash2 = murmurhash3(input, seed);
    try std.testing.expect(hash == hash2);
}

test "bounded hash generation" {
    const user_id = "test_user";
    const target_id = "test_target";
    
    const bounded_hash = generateBoundedHashes(user_id, target_id);
    
    // Hash values should be between 0 and 1
    try std.testing.expect(bounded_hash.rollout_hash >= 0.0 and bounded_hash.rollout_hash <= 1.0);
    try std.testing.expect(bounded_hash.bucketing_hash >= 0.0 and bounded_hash.bucketing_hash <= 1.0);
}

test "consistent hashing" {
    const user_id = "consistent_user";
    const target_id = "consistent_target";
    
    const hash1 = generateBoundedHashes(user_id, target_id);
    const hash2 = generateBoundedHashes(user_id, target_id);
    
    // Same inputs should produce same outputs
    try std.testing.expect(hash1.rollout_hash == hash2.rollout_hash);
    try std.testing.expect(hash1.bucketing_hash == hash2.bucketing_hash);
}